File: src/openweather.ts
import * as logger from "./logger";
import * as request from "./request";

export interface Config {
  apiKey: string;
  requester?: request.Getter;
  logger: logger.Logger;
}

export interface Openweather {
  forecast: (config: Config, city: string, country: string) => Promise<Probe[]>;
}

export interface Probe {
  dt: number;
  date: Date;
  temperature: number;
  humidity: number;
}

interface OpenweatherProbeAPI {
  dt: number;
  main: {
    temp: number;
    humidity: number;
  };
}

const probeFromAPI = (rawProbe: OpenweatherProbeAPI): Probe => {
  const date = new Date();
  date.setTime(rawProbe.dt * 1000);

  return {
    date: date,
    dt: rawProbe.dt,
    temperature: rawProbe.main.temp,
    humidity: rawProbe.main.humidity
  };
};

export const forecast = (
  config: Config,
  city: string,
  country: string
): Promise<Probe[]> => {
  const requester = config.requester ? config.requester : request;

  const query = {
    q: `${city},${country}`,
    units: "metric",
    appid: config.apiKey
  };

  config.logger.debug("start searching on openweather");

  return requester
    .get({
      uri: "https://api.openweathermap.org/data/2.5/forecast",
      qs: query
    })
    .then(body => JSON.parse(body))
    .then(response => {
      config.logger.debug("response received from openweather");

      return response;
    })
    .then(response => response.list.map(probeFromAPI));
};

File: internal/openweather/openweather.go
package openweather

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

type HTTPDoer interface {
	Do(*http.Request) (*http.Response, error)
}

type OpenWeather struct {
	APIKey   string
	Country  string
	ZipCode  string
	HTTPDoer HTTPDoer
}

type Probe struct {
	TempCelsius        float64
	HumidityPercentage float64
}

func New(APIKey string, country string, zipCode string) *OpenWeather {
	return &OpenWeather{
		APIKey:   APIKey,
		Country:  country,
		ZipCode:  zipCode,
		HTTPDoer: http.DefaultClient,
	}
}

type fivedayforecastResponse struct {
	HTTPCode string      `json:"cod"`
	Message  interface{} `json:"message"`
	Probes   []struct {
		Timestamp int64 `json:"dt"`
		Main      struct {
			Temp     float64 `json:"temp"`
			Humidity float64 `json:"humidity"`
		} `json:"main"`
	} `json:"list"`
}

func (o *OpenWeather) Get(ctx context.Context, from time.Time, to time.Time) ([]Probe, error) {
	request, _ := http.NewRequestWithContext(ctx, "GET", o.fiveDayForecastURL(), nil)
	response, err := o.HTTPDoer.Do(request)
	if err != nil {
		return nil, fmt.Errorf("can't request openweather API: %w", err)
	}

	var allprobes fivedayforecastResponse
	decoder := json.NewDecoder(response.Body)
	if err := decoder.Decode(&allprobes); err != nil {
		return nil, fmt.Errorf("can't decode openweather response: %w", err)
	}

	if allprobes.HTTPCode != "200" {
		return nil, fmt.Errorf("can't get valid openweather response: %#+v", allprobes)
	}

	var nightprobes []Probe
	for _, probe := range allprobes.Probes {
		probeTime := time.Unix(probe.Timestamp, 0)
		if probeTime.After(from) && probeTime.Before(to) {
			nightprobes = append(nightprobes, Probe{
				TempCelsius:        probe.Main.Temp,
				HumidityPercentage: probe.Main.Humidity,
			})
		}
	}

	return nightprobes, nil
}

func (o *OpenWeather) fiveDayForecastURL() string {
	forecast, _ := url.Parse("https://api.openweathermap.org/data/2.5/forecast")

	querystring := forecast.Query()
	querystring.Set("appid", o.APIKey)
	querystring.Set("zip", fmt.Sprintf("%s,%s", o.ZipCode, o.Country))
	querystring.Set("units", "metric")
	forecast.RawQuery = querystring.Encode()

	return forecast.String()
}
